nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py

```diff
@@ -0,0 +1,423 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import re
+from operator import itemgetter
+from typing import Any, Dict, Optional
+
+import torch
+import torch.nn as nn
+from omegaconf.dictconfig import DictConfig
+from pytorch_lightning.trainer.trainer import Trainer
+
+from nemo.collections.nlp.data.language_modeling.megatron.data_samplers import (
+    MegatronPretrainingRandomSampler,
+    MegatronPretrainingSampler,
+)
+from nemo.collections.nlp.models.language_modeling.megatron.tokens_encoder_decoder import TokensEncoderDecoderModule
+from nemo.collections.nlp.models.language_modeling.megatron_base_model import MegatronBaseModel
+from nemo.collections.nlp.modules.common.megatron.clip_grads import clip_grad_norm_fp32
+from nemo.collections.nlp.modules.common.megatron.utils import average_losses_across_data_parallel_group
+from nemo.collections.nlp.modules.common.tokenizer_utils import get_nmt_tokenizer
+from nemo.utils import AppState, logging
+
+try:
+    from apex.transformer import parallel_state, tensor_parallel
+
+    HAVE_APEX = True
+except (ImportError, ModuleNotFoundError):
+    HAVE_APEX = False
+
+
+__all__ = ["MegatronLMEncoderDecoderModule"]
+
+
+class MegatronLMEncoderDecoderModule(MegatronBaseModel):
+    """
+    Megatron encoder-decoder base class
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer=trainer)
+
+        self.tokenizer = get_nmt_tokenizer(
```

Comment:
@michalivne Would it be possible to move the tokenizer creation to the `MegatronT5Model` class? As currently implemented, `MegatronLMEncoderDecoderModule` is specific to NMT because of `get_nmt_tokenizer`), and it appears to be the only thing imposing this limitation -- the rest of the code is generally applicable. Because it is written in such a nice general style, I'd prefer to just subclass `MegatronLMEncoderDecoderModule`, but the tokenizer part will be a challenge. 

Seems like this could be true for other applications too, as I see that additional tokenizer getter functions exist.

Reply:
The other option, to implement a new `__init__` function, will also not work for MegaMolBART because the model is dependent upon things initialized in `MegatronBaseModel` and `NLPModel`.

Reply:
Good idea. Will provide an easy override.

Reply:
Done. Great feedback @mlgill! You can now override `_build_tokenizer` method in `MegatronLMEncoderDecoderModule`

Reply:
Looks great, thank you!